infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationDiscoveryService.java

```diff
@@ -494,41 +514,155 @@ private void registerToLogReplicationLock() {
+            // Replication Context contains the topology which must be updated if it has changed
+            replicationManager.setContext(replicationContext);
+        }
+
+        initReplicationStatusForRemoteClusters(topologyDescriptor.getRemoteSinkClusters().keySet(), true);
+        initReplicationStatusForRemoteClusters(topologyDescriptor.getRemoteSourceClusters().keySet(), false);
+
+        Set<ClusterDescriptor> connectionEndpoints = new HashSet<>(fetchConnectionEndpoints().values());
+        Set<ClusterDescriptor> remoteSinkClusters = new HashSet<>(topologyDescriptor.getRemoteSinkClusters().values());
+        Set<ClusterDescriptor> remoteSourceClusters = new HashSet<>(topologyDescriptor.getRemoteSourceClusters().values());
+
+        log.debug("remoteSourceCluster: {}, remoteSinkCluster: {} , connectionEndpoints: {}",
+                topologyDescriptor.getRemoteSourceClusters().values(), topologyDescriptor.getRemoteSinkClusters().values(),
+                fetchConnectionEndpoints().values());
+
+        // check if local cluster is a connectionReceiver. This is used to start the server components
+        remoteSinkClusters.removeAll(connectionEndpoints);
+        remoteSourceClusters.removeAll(connectionEndpoints);
+        boolean isConnectionReceiver = !remoteSinkClusters.isEmpty() || !remoteSourceClusters.isEmpty();
+
+        setupConnectionReceiversAndSinkServers(isConnectionReceiver, !topologyDescriptor.getRemoteSourceClusters().isEmpty());
+
+        // create connection endpoint routers to pass to the interClusterServerNode.
+        // The incoming msgs will be routed to the appropriate router in the transport layer adapter.
+        Map<ReplicationSession, LogReplicationSourceServerRouter> receivedSessionToSourceServer = new HashMap<>();
+        Map<ReplicationSession, LogReplicationSinkServerRouter> receivedSessionToSinkServer = new HashMap<>();
+
+        if(!remoteSinkClusters.isEmpty()) {
+            startSourceComponents(connectionEndpoints, receivedSessionToSourceServer);
+        }
+
+        if (!remoteSourceClusters.isEmpty()) {
+            startSinkComponents(connectionEndpoints, receivedSessionToSinkServer);
+        }
+
+        if(isConnectionReceiver) {
+            interClusterServerNode.setRouterAndStartServer(receivedSessionToSourceServer, receivedSessionToSinkServer);
+        }
+
+        // connect to endpoints
+        replicationManager.startConnection(connectionEndpoints, remoteClusterIdToReplicationSession, serverMap);
+
+        lockAcquireSample = recordLockAcquire(localClusterDescriptor.getRole());
+        processCountOnLockAcquire(localClusterDescriptor.getRole());
+    }
+
+
+    private void setupConnectionReceiversAndSinkServers(boolean isConnectionReceiver, boolean isSink) {
+        if(!isConnectionReceiver && !isSink) {
+            return;
+        }
+
+        LogReplicationServer lrServer = new LogReplicationServer(serverContext, localNodeId, replicationConfigManager,
```

Comment:
why should the server be started if a node is not a receiver?

Reply:
I have tried to keep the various msg processing (leadership/negotiation/replication/Acks) logic the same
- SINK will need LRServer irrespective of if its a connection starter/receiver. Because the handling of negotiation and replication msgs are in this class.
- SOURCE will need LRServer when its the connection receiver, as I am reusing the LR_LEADERSHIP query handling that in the class.

Also, the LRServer is instantiated only once. So if a node is both the SOURCE and SINK, we use the same object